src/Compilers/Core/Portable/Emit/EditAndContinue/DeltaMetadataWriter.cs

```diff
@@ -784,6 +810,154 @@ private void PopulateEncLogTableParameters()
+            // in the original metadata for this algorithm to work.
+            var attributeTargetsEmitted = _customAttributeTargets.OrderBy(att => CodedIndex.HasCustomAttribute(att)).ToList();
+
+            // The first phase of this is to iterate through the (potentially huge) custom attributes from the original
+            // compilation, and issue updates for existing rows as we go.
+            // This call mutates the list passed in, removing anything it emits a log entry for.
+            PopulateEncLogTableCustomAttributesFromOriginalMetadata(attributeTargetsEmitted);
+
+            // The second phase is to go through the rows emitted for each delta and issue updates to them, and also
+            // additions for new attributes, from whatever is left in the list.
+            PopulateEncLogTableCustomAttributesFromDeltas(attributeTargetsEmitted);
+        }
+
+        private void PopulateEncLogTableCustomAttributesFromOriginalMetadata(List<EntityHandle> attributeTargetsEmitted)
+        {
+            var metadataReader = _previousGeneration.OriginalMetadata.MetadataReader;
+
+            var rowId = 1;
+            var emittedIndex = 0;
+
+            // We do this by also iterating through the emitted attributes at the same time, since they're sorted the same.
+            foreach (var customAttributeHandle in metadataReader.CustomAttributes)
+            {
+                // Data from the existing CA table
+                var parent = metadataReader.GetCustomAttribute(customAttributeHandle).Parent;
+                var parentCodedIndex = CodedIndex.HasCustomAttribute(parent);
+
+                // Data from the CA table delta we are emitting
+                var emitted = attributeTargetsEmitted[emittedIndex];
+                var emittedCodedIndex = CodedIndex.HasCustomAttribute(emitted);
+
+                // If the existing data is ahead of the emitted data then it means the current emitted attribute
+                // must be for something that wasn't in the original compilation so we'll deal with it later
+                if (parentCodedIndex > emittedCodedIndex)
+                {
+                    // Skip this current attribute because we didn't find it
+                    emittedIndex++;
+                    if (emittedIndex == attributeTargetsEmitted.Count)
+                    {
+                        break;
+                    }
+                    emitted = attributeTargetsEmitted[emittedIndex];
+                }
+
+                if (parent == emitted)
+                {
+                    // If we found the attribute we want then add an EncLog table record for it
+                    AddEncLogCustomAttributeEntry(rowId);
+                    // Remove the attribute from our list because we've already dealt with it
+                    attributeTargetsEmitted.RemoveAt(emittedIndex);
```

Comment:
>attributeTargetsEmitted.RemoveAt(emittedIndex); [](http://example.com/codeflow?start=20&length=47)

wouldn't it be cleaner to add the ones that were not matched to another list instead of removing them from the current list?